state/remoteapplication.go

```diff
@@ -498,6 +499,65 @@ func (s *RemoteApplication) SetStatus(info status.StatusInfo) error {
 	})
 }
 
+// TerminateOperation return a ModelOperation that will terminate this
+// remote application when applied, ensuring that all units have left
+// scope as well.
+func (s *RemoteApplication) TerminateOperation(message string) ModelOperation {
+	return &terminateRemoteApplicationOperation{
+		app: s,
+		doc: statusDoc{
+			Status:     status.Terminated,
+			StatusInfo: message,
+			Updated:    s.st.clock().Now().UnixNano(),
+		},
+	}
+}
+
+type terminateRemoteApplicationOperation struct {
+	app *RemoteApplication
+	doc statusDoc
+}
+
+// Build is part of ModelOperation.
+func (op *terminateRemoteApplicationOperation) Build(attempt int) ([]txn.Op, error) {
+	ops, err := statusSetOps(op.app.st.db(), op.doc, op.app.globalKey())
+	if err != nil {
+		return nil, errors.Annotate(err, "setting status")
+	}
+	name := op.app.Name()
+	logger.Debugf("leaving scope on all %q relation units", name)
+	rels, err := op.app.Relations()
+	if err != nil {
+		return nil, errors.Annotatef(err, "getting relations for %q", name)
+	}
+	for _, rel := range rels {
+		remoteUnits, err := rel.AllRemoteUnits(name)
+		if err != nil {
+			return nil, errors.Annotatef(err, "getting remote units for %q", name)
+		}
+		for _, ru := range remoteUnits {
+			leaveOperation := ru.LeaveScopeOperation(false)
+			// Don't pass attempt down since the relationunit is
+			// always fresh.
+			leaveOps, err := leaveOperation.Build(0)
```

Comment:
I think it's safer to pass attempt because if the LeaveScopeOperation changes implementation, it will be robust. Passing 0 always isn't robust to such changes.

Reply:
Yeah, that makes sense - it seems wasteful since they were just loaded, but you're right that it could be doing something else with attempt.